iceoryx_utils/include/iceoryx_utils/internal/units/duration.inl

```diff
@@ -21,261 +21,410 @@ namespace iox
+            m_seconds = std::numeric_limits<uint64_t>::max();
+            m_nanoseconds = NANOSECS_PER_SEC - 1U;
+        }
+        else
+        {
+            m_seconds += additionalSeconds;
+            m_nanoseconds = m_nanoseconds % NANOSECS_PER_SEC;
+        }
+    }
 }
 
-inline constexpr Duration::Duration(const struct timespec& value)
-    : Duration(static_cast<long double>(value.tv_sec) + static_cast<long double>(value.tv_nsec) / 1000000000.0)
+inline constexpr Duration::Duration(const struct timeval& value) noexcept
+    : Duration(static_cast<uint64_t>(value.tv_sec), static_cast<uint32_t>(value.tv_usec) * NANOSECS_PER_MICROSEC)
 {
 }
 
-inline constexpr Duration::Duration(const struct itimerspec& value)
-    : Duration(value.it_interval)
+inline constexpr Duration::Duration(const struct timespec& value) noexcept
+    : Duration(static_cast<uint64_t>(value.tv_sec), static_cast<uint32_t>(value.tv_nsec))
 {
 }
 
-inline constexpr Duration::Duration(const std::chrono::milliseconds& value)
-    : Duration(static_cast<long double>(value.count()) / 1000.0)
+inline constexpr Duration::Duration(const struct itimerspec& value) noexcept
+    : Duration(value.it_interval)
 {
 }
 
-inline constexpr Duration::Duration(const std::chrono::nanoseconds& value)
-    : Duration(static_cast<long double>(value.count()) / 1000000000.0)
+inline constexpr Duration::Duration(const std::chrono::milliseconds& value) noexcept
 {
+    *this = Duration::milliseconds(value.count());
 }
 
-inline Duration& Duration::operator=(const std::chrono::milliseconds& right)
+inline constexpr Duration::Duration(const std::chrono::nanoseconds& value) noexcept
 {
-    durationInSeconds = static_cast<long double>(right.count()) / 1000.0;
-    return *this;
+    *this = Duration::nanoseconds(value.count());
 }
 
-inline constexpr Duration::Duration(const long double durationInSeconds)
-    : durationInSeconds(durationInSeconds >= 0.0 ? durationInSeconds : 0.0)
+inline Duration& Duration::operator=(const std::chrono::milliseconds& rhs) noexcept
 {
+    *this = Duration(rhs);
+    return *this;
 }
 
 template <typename T>
-inline constexpr T Duration::nanoSeconds() const
+inline constexpr T Duration::nanoSeconds() const noexcept
 {
-    return static_cast<T>(durationInSeconds * 1000000000.0);
+    static_assert(std::numeric_limits<T>::is_integer, "only integer are supported");
+
+    constexpr uint64_t MAX_SECONDS_BEFORE_OVERFLOW{std::numeric_limits<T>::max() / NANOSECS_PER_SEC};
+    constexpr uint64_t MAX_NANOSECONDS_BEFORE_OVERFLOW{std::numeric_limits<T>::max() % NANOSECS_PER_SEC};
+    constexpr Duration MAX_DURATION_BEFORE_OVERFLOW{MAX_SECONDS_BEFORE_OVERFLOW, MAX_NANOSECONDS_BEFORE_OVERFLOW};
+
+    if (*this > MAX_DURATION_BEFORE_OVERFLOW)
+    {
+        std::clog << __PRETTY_FUNCTION__ << ": Result of conversion would overflow, clamping to max value!"
+                  << std::endl;
+        return std::numeric_limits<T>::max();
+    }
```

Comment:
@elfenpiff this is the price we are paying for the overflow detection. I think it's not too bad and could be disabled for release builds if necessary